### ISSUE TITLE
Elastic Inference mixes up models and gives incorrect results in pytorch

### ISSUE THREAD

**Issue by jamt9000** [aws/sagemaker-python-sdk#1959]
**Describe the bug**
When you have two models of the same architecture (or possibly just the same class name?), EI will return the results for one model when you call forward on a different model. This causes predictions to be completely erroneous when deploying models with EI.

**To reproduce**

This is a minimal repro with a torchvision resnet - but the problem happens even when the architectures aren't exactly the same, such as having different numbers of outputs.

```.py
import torch, torchvision
import PIL
from torchvision import transforms
from PIL import Image

torch.manual_seed(1)

im = torch.randn(1,3,224,224)

USE_EI = True

if USE_EI:
    jit_opts = (True, {'target_device': 'eia:0'})
else:
    jit_opts = (True,)

# Setup two Resnet50 models, one with pretrained weights, the other
# initialised from scratch
model = torchvision.models.resnet50(pretrained=True).eval()
with torch.jit.optimized_execution(*jit_opts):
  model = torch.jit.trace(model, im)

model2 = torchvision.models.resnet50(pretrained=False).eval()
with torch.jit.optimized_execution(*jit_opts):
  model2 = torch.jit.trace(model2, im)

# Run both models with EI. The results ought to be different
with torch.no_grad():
  with torch.jit.optimized_execution(*jit_opts):
    probs = model(im)

print('probs1', probs[0][:20])

with torch.no_grad():
  with torch.jit.optimized_execution(*jit_opts):
    probs2 = model2(im)

print('probs2', probs2[0][:20])
```

**Expected behavior**

Expected output without EI - the model outputs are different

```
python eirepro.py
probs1 tensor([-0.5483,  0.0222,  0.0213, -0.8796,  0.5466, -1.3745, -0.0786, -0.3982,
         0.4778, -1.4499, -0.1451,  1.2918,  0.0519, -0.1024, -1.0931, -0.7273,
        -1.1181, -0.7758,  0.7030,  0.4041])
probs2 tensor([-19.1023,  -9.6511,  -7.8852, -26.7294,  -2.1230,   0.4408,  -8.6310,
         22.3135,   1.0720, -16.9864, -11.1711, -26.5716,  -7.0350,   0.9620,
         16.8690,   6.6240,   5.5423,  20.1270,  38.7582,  34.2791])
```

Actual output with EI:

```
/home/ubuntu/anaconda3/envs/amazonei_pytorch_p36/lib/python3.6/site-packages/torch/jit/__init__.py:1019: TracerWarning: Output nr 1. of the traced function does not match the corresponding output of the Python function. Detailed error:
Not within tolerance rtol=1e-05 atol=1e-05 at input[0, 970] (-1.1768685579299927 vs. -67.72380828857422) and 999 other locations (100.00%)
  check_tolerance, _force_outplace, True, _module_class)
probs1 tensor([-0.5483,  0.0222,  0.0213, -0.8796,  0.5466, -1.3745, -0.0786, -0.3982,
         0.4778, -1.4499, -0.1451,  1.2918,  0.0519, -0.1024, -1.0931, -0.7273,
        -1.1181, -0.7758,  0.7030,  0.4041])
probs2 tensor([-0.5483,  0.0222,  0.0213, -0.8796,  0.5466, -1.3745, -0.0786, -0.3982,
         0.4778, -1.4499, -0.1451,  1.2918,  0.0519, -0.1024, -1.0931, -0.7273,
        -1.1181, -0.7758,  0.7030,  0.4041])
```

**System information**:

amazonei_pytorch_p36 with provided pytorch 1.3.1

**Comment by ChuyangDeng** [comment]
Hi @jamt9000, could you be more specific about the container version you are having issue with? Our PyTorch inference container uses torchserve starting from version 1.6. With version 1.3.1, we uses multi-model-server instead.

When you mentioned 2 models, are you using Multi-Model endpoint?

**Comment by rucha3** [comment]
I am facing the exact same issue. I am using the container `pytorch-inference-eia:1.3.1-cpu-py36-ubuntu16.04` for inference.

I am actually having an ensemble of 3 models and I am getting prediction from each of them in my code.
Each model is a fine-tuned distilBERT model and has different number of output labels.

When I run this container without EI, everything works fine. All 3 models produce different outputs of different shapes.

But when I use EI i.e. add `{'target_device': 'eia:0'}` to `torch.jit.optimized_execution`, then all the models produce the same output as the first one thats executed. Each of the models is used in its own separate `torch.jit.optimized_execution` context.

My `predict_fn` method looks something like this:
```
def predict_fn(input_object, models):
    input_ids = input_object["input_ids"]
    attention_masks = input_object["attention_mask"]
    models = [model1, model2, model3]
    outputs = []
    for i, model in enumerate(models):
        with torch.no_grad():
            if EI_ENABLED:
                with torch.jit.optimized_execution(True, {"target_device": "eia:0"}): 
                    print("Using EI")
                    output = model(input_ids, attention_masks)[0]
                else:
                    with torch.jit.optimized_execution(True):
                    output = model(input_ids, attention_masks)[0]
        
        print(f"Output of {i}th model = {output}")
        print(f"Shape of output of {i}th model = {output.shape}")
        outputs.append(output)
    # remaining ensemble code...

```

**Comment by jamt9000** [comment]
> Hi @jamt9000, could you be more specific about the container version you are having issue with? Our PyTorch inference container uses torchserve starting from version 1.6. With version 1.3.1, we uses multi-model-server instead.
> 
> When you mentioned 2 models, are you using Multi-Model endpoint?

For the repro I gave I was just using the Deep Learning AMI with no container - but this happens for any use of Elastic Inference, as shown above by @rucha3. The two models are just instantiated with standard python/pytorch.

**Comment by mohanasudhan** [comment]
@jamt9000 Thanks for reporting this issue and apologies for delay in the response. Please refer to Elastic Inference document for recommendation https://docs.aws.amazon.com/sagemaker/latest/dg/ei.html

Q: We currently use Amazon Elastic Inference (EI) accelerators. Will we be able to continue using them after April 15, 2023?

Yes, you will be able use Amazon EI accelerators. We recommend that you migrate your current ML Inference workloads running on Amazon EI to other hardware accelerator options at your earliest convenience.

Recommend going over the migration steps https://docs.aws.amazon.com/sagemaker/latest/dg/ei.html#ei-migration

Please reopen the issue if you see any issue with migration.
